### PR TITLE
Unify datasource configuration

### PR DESCRIPTION
**Pull request by frascuchon**
This  PR implements definition described in [this pitch](https://gitlab.com/recognai-team/team/issues/42), keeping completely backward compatibility for old configurations

### CODE REVIEW

File: src/biome/data/sources/datasource.py [hunk 1]
```diff
@@ -28,46 +26,74 @@ class DataSource:
...
+    # maps the supported formats to the corresponding "source readers"
 
     def __init__(
-        self, format: str, mapping: Dict[str, Union[List[str], str]] = None, **kwargs
+        self,
+        source: Optional[Union[str, List[str]]] = None,
+        attributes: Optional[Dict[str, Any]] = None,
+        mapping: Optional[Dict[str, Union[List[str], str]]] = None,
+        format: Optional[str] = None,
+        **kwargs,
```
Comment (dcfidalgo) [inline src/biome/data/sources/datasource.py:65]: see general comment

File: tests/data/sources/test_datasource.py [hunk 2]
```diff
@@ -24,31 +26,54 @@ def ds_parser(**kwargs):
             )
 
         DataSource.add_supported_format("new-format", ds_parser)
-        DataSource.add_supported_format("new-format", ds_parser)
-        ds = DataSource(format="new-format")
-        self.assertFalse(ds.to_dataframe().columns is None)
+
+        for ds in [DataSource(format="new-format"), DataSource(source="new-format")]:
```
Comment (dcfidalgo) [inline tests/data/sources/test_datasource.py:30]: I think we should not test deprecated behavior to free us from overload. That is, no need to test `DataSource(format="new-format")`.

File: tests/data/sources/test_datasource.py [hunk 3]
```diff
@@ -24,31 +26,54 @@ def ds_parser(**kwargs):
...
-            path=os.path.join(FILES_PATH, "dataset_source.jsonl"),
-        )
+        the_mapping = {"label": "overall", "tokens": "summary"}
 
-        df = ds.to_mapped_dataframe()
+        for ds in [
+            DataSource(
+                format="json",
+                mapping=the_mapping,
+                path=os.path.join(FILES_PATH, "dataset_source.jsonl"),
```
Comment (dcfidalgo) [inline tests/data/sources/test_datasource.py:40]: See comment above, no need to test deprecated behavior.

File: tests/data/sources/test_datasource.py [hunk 4]
```diff
@@ -24,31 +26,54 @@ def ds_parser(**kwargs):
...
 
-        self.assertIn("label", bag)
-        self.assertIn("tokens", bag)
+            self.assertIn("label", bag)
+            self.assertIn("tokens", bag)
 
     def test_no_mapping(self):
+        for ds in [
+            DataSource(
+                format="json", path=os.path.join(FILES_PATH, "dataset_source.jsonl")
```
Comment (dcfidalgo) [inline tests/data/sources/test_datasource.py:60]: see comment above

**Review by dcfidalgo — COMMENTED** [review]
Some inline comments and a general comment: I would rather move the whole backward compatibility thing from the `init` to the `.from_yaml` method. This makes sure that we still support old yaml files, but forces direct calls to the `DataSource` class to use the new format. I guess at some point (for a v1.0 version at least), we want to drop deprecated things, and if we do not address direct calls now, we have to do it in the future. For me the best way would be to put everything in a method or class that we can simply delete when we decide to drop backward compatibility.

**Comment by frascuchon** [comment]
> Some inline comments and a general comment: I would rather move the whole backward compatibility thing from the `init` to the `.from_yaml` method. This makes sure that we still support old yaml files, but forces direct calls to the `DataSource` class to use the new format. I guess at some point (for a v1.0 version at least), we want to drop deprecated things, and if we do not address direct calls now, we have to do it in the future. For me the best way would be to put everything in a method or class that we can simply delete when we decide to drop backward compatibility.

I'm agree with you but, since we are working with this library without using tags, I prefer completely keep backward compatibility warning the use of deprecated stuffs and progressively migrate the rest of the modules. The changes you mention make crash at least the `biome-classifer` and `biome-extraction` modules. So, i prefer integrate this into master without break anything

**Review by dvsrepo — APPROVED** [review]

**Pull request merged**